flang/lib/Optimizer/Analysis/AliasAnalysis.cpp

```diff
@@ -159,21 +193,56 @@ AliasResult AliasAnalysis::alias(Value lhs, Value rhs) {
     src2->attributes.set(Attribute::Target);
   }
 
-  // Dummy TARGET/POINTER argument may alias with a global TARGET/POINTER.
+  // Two TARGET/POINTERs may alias.
   if (src1->isTargetOrPointer() && src2->isTargetOrPointer() &&
       src1->isData() == src2->isData()) {
     LLVM_DEBUG(llvm::dbgs() << "  aliasing because of target or pointer\n");
     return AliasResult::MayAlias;
   }
 
-  // Box for POINTER component inside an object of a derived type
-  // may alias box of a POINTER object, as well as boxes for POINTER
-  // components inside two objects of derived types may alias.
-  if ((src1->isRecordWithPointerComponent() && src2->isTargetOrPointer()) ||
-      (src2->isRecordWithPointerComponent() && src1->isTargetOrPointer()) ||
-      (src1->isRecordWithPointerComponent() &&
-       src2->isRecordWithPointerComponent())) {
-    LLVM_DEBUG(llvm::dbgs() << "  aliasing because of pointer components\n");
+  // A pointer dummy arg (but not a pointer component of a dummy arg) may alias
+  // a pointer component and thus the associated composite.  That composite
+  // might be a global or another dummy arg.  This is an example of the global
+  // composite case:
+  //
+  // module m
+  //   type t
+  //      real, pointer :: p
+  //   end type
+  //   type(t) :: a
+  //   type(t) :: b
+  // contains
+  //   subroutine test(p)
+  //     real, pointer :: p
+  //     p = 42
+  //     a = b
+  //     print *, p
+  //   end subroutine
+  // end module
+  // program
+  //   use m
+  //   real, target :: x1 = 1
+  //   real, target :: x2 = 2
+  //   a%p => x1
+  //   b%p => x2
+  //   call test(a%p)
+  // end
+  //
+  // The dummy argument p is an alias for a%p, even for the purposes of pointer
+  // association during the assignment a = b.  Thus, the program should print 2.
+  if ((isRecordWithPointerComponent(val1->getType()) &&
+       src1->kind != SourceKind::Allocate &&
+       src2->kind == SourceKind::Argument &&
+       src2->attributes.test(Attribute::Pointer) && !src2->isData() &&
+       !isRecordWithPointerComponent(src2->valueType)) ||
```

Comment:
Why is the test "!isRecordWithPointerComponent(src2->valueType)" needed?
It seems either wrong or useless to me (the pointer component could itself point towards a derived type with pointer component).

Reply:
Thanks for reviewing.  

The `if` preceding this one uses `isTargetOrPointer` and covers aliasing among addresses *in* pointers (what they point towards).  The `if` here has the `!src2->isData()` condition and so examines only the address *of* the pointer.

Specifically, the `if` here checks for aliasing with the address of a pointer that is a dummy argument.  The condition `!isRecordWithPointerComponent(src2->valueType)` prevents it from mistaking a pointer component of a dummy argument as a pointer dummy argument.  For example, [the test alias-analysis-9.fir has a check](https://github.com/jdenny-ornl/llvm-project/blob/1e5fdd49f4c8558f5181b3bdb2d38002e0b34630/flang/test/Analysis/AliasAnalysis/alias-analysis-9.fir#L31-L34) whose result would change without that condition:

```
// module m
// type t
//  type(t), pointer :: next
//  integer :: i
// end type
// contains
// subroutine foo(x, y)
//   type(t) :: x, y
//   integer :: i1, i2
//   i1 = x%next%i
//   x = y
//   i2 = x%next%i
// end subroutine
// end module

// CHECK-DAG: y#0 <-> xnext1#0: NoAlias
```

`y#0` is the address of `y`.  `xnext1#0` is the address of `x%next`. Without the above condition, AliasAnalysis would treat `xnext1#0` as the address of a pointer dummy argument, and the result above would become MayAlias.

Reply:
Thanks for the explanation, while I understand the goal, the condition is still wrong in my opinion.

> the result above would become MayAlias

That would be a valid if the dummies had the TARGET attribute as per 15.5.2.14 (3) (b).

Here is an illustration where "a" and "c%p" should be considered to potentially alias (which is not enforced properly because of this condition). :
```
module m
  type t 
     real, pointer :: p
  end type
  type(t) :: a
  type(t) :: b
contains
subroutine test(c)
  type(t), target :: c
  c%p = 42
  a = b
  print *, c%p
end subroutine 
end module
    
  use m
  real, target :: x1 = 1
  real, target :: x2 = 2
  a%p => x1
  b%p => x2
  call test(a)
end 
```

To pass this at the FIR level, the `src2->attributes.test(Attribute::Pointer)` should also be removed because the attributes where not gathered on the hlfir.designate which does not exist anymore in FIR (replaced by fir.coordinate). This makes some sense anyway since the source (the dummy) in itself does not really has that attribute anyway.

Reply:
I find it confusing that the `target` attribute on `c` is relevant in your example given that there are no pointers or other targets with the type of `c`.

Here is Fortran 2023 sec. 15.5.2.14, 1 (3) (b):

> 1 While an entity is associated with a dummy argument, the following restrictions hold.
>
> (3) Action that affects the value of the entity or any subobject of it shall be taken only through the dummy argument unless
>
> (b) the dummy argument is a scalar, assumed-shape, or assumed-rank object, and has the TARGET attribute but not the INTENT (IN) or CONTIGUOUS attributes, and the actual argument is a target other than a coindexed object or an array section with a vector subscript,

`c` is the dummy argument.  `a` is the actual argument.  But the above requires that "the actual argument is a target", which 3.141 defines as:

> entity that is pointer associated with a pointer (3.104), entity on the right-hand-side of a pointer assignment statement, or entity with the TARGET attribute

I don't see how `a` is a target in your example, so I don't see how your example is valid Fortran.  I think we can fix your example by adding the `target` attribute to `a`.  Do you agree?

It's a surprise to me to that the `target` attribute can be used to permit aliasing with other targets via dummy arguments.  I thought the sole purpose of the `target` attribute was to enable an entity to be the target of a pointer.

I want to make sure we're on the same page so far before I start working on fixing this case.

Reply:
`a` in my example does not need target because it is not a dummy argument, so `15.5.2.14` does not restrict the effects on it. `15.5.2.14` restrict the effects on `c`, so it only matters that `c` has this attribute to allow `c` to be effected by changes on `a`.

I agree this is not a very well known aspect, but it is what is, and other compiler respect it (if you do A= B with two TARGET assumed-shape, you should observe that all compiler will cope with the case where A an B alias, even f not POINTERs are involved).

Reply:
> `a` in my example does not need target because it is not a dummy argument, so `15.5.2.14` does not restrict the effects on it. `15.5.2.14` restrict the effects on `c`, so it only matters that `c` has this attribute to allow `c` to be effected by changes on `a`.

Thanks, but would you please help me to understand how the spec says that?  Or maybe the spec is stricter than defacto behavior?  In particular, when the dummy argument (`c`) has the TARGET attribute, one condition in the spec to permit aliasing is that the _actual argument_ (`a`) is a target.  Given the definition of "target" that I quoted, how is `a` a target in your example?
> 
> I agree this is not a very well known aspect, but it is what is, and other compiler respect it (if you do A= B with two TARGET assumed-shape, you should observe that all compiler will cope with the case where A an B alias, even f not POINTERs are involved).

I assume this doesn't refer to the `a` and `b` in your example given that neither have TARGET.  In this new example, are A and B both dummy arguments, both with TARGET?

Reply:
Thank you for correcting me here, yes TARGET is needed on `a` for my example to be valid after reading that again. I verified with a few compilers that the behavior in `A=B` differs depending on A attribute in the following program (prints 1.0 if A has the attribute, prints N otherwise):

```
module m
integer, parameter :: n = 10000 
real, target :: a(n) = [(i, i=1,n)]
contains

subroutine test_aliasing(b)
  real, target :: b(:)
  a = b
end subroutine
end module

  use m
  call test_aliasing(a(n:1:-1))
  print *, a(n)
end
```

And yes, that was different A and B, sorry for the confusion.

Reply:
Thanks for confirming.  I'll work on a fix.

Reply:
I [added a fix](https://github.com/llvm/llvm-project/pull/94242/commits/a61be3a0d7202da4ccd9a3b38dc68dc3f119c2ce) for that.